### PR TITLE
[vulkan] Improve Vulkan RHI impl with lower overhead internal implementations

### PR DESCRIPTION
**Pull request by bobcao3**
Also added partial RHI support for Vulkan Dynamic Rendering, but there are some interactions and bugs to figure out between this and ImGUI, so the device creator will not actively enable this feature for now.

### CODE REVIEW

File: taichi/rhi/impl_support.h [hunk 1]
```diff
@@ -89,5 +91,37 @@ struct BidirMap {
   }
 };
 
+// A synchronized list of objects that is pointer stable & reuse objects
+// It does not mark objects as used, and it does not free objects (destructor is
+// not called)
+template <class T>
+struct SyncedPtrStableObjectList {
+  std::mutex lock;
```
Comment (k-ye) [inline taichi/rhi/impl_support.h:99]: nit: make these members private and put them to the end?

File: taichi/rhi/device.h [hunk 2]
```diff
@@ -17,6 +17,7 @@ enum class RhiResult {
   error = -1,
   invalid_usage = -2,
   not_supported = -3,
+  out_of_memory = -4,
```
Comment (PENGUINLIONG) [inline taichi/rhi/device.h:20]: 🤩

File: taichi/rhi/vulkan/vulkan_device.cpp [hunk 3]
```diff
@@ -827,8 +858,11 @@ void VulkanCommandList::bind_pipeline(Pipeline *p) {
     return;
 
   if (pipeline->is_graphics()) {
```
Comment (PENGUINLIONG) [inline taichi/rhi/vulkan/vulkan_device.cpp:860]: i wonder if we will have an way to opt out graphics support in the distributed runtime library in the future?

File: taichi/rhi/impl_support.h [hunk 4]
```diff
@@ -89,5 +91,39 @@ struct BidirMap {
   }
 };
 
+// A synchronized list of objects that is pointer stable & reuse objects
+// It does not mark objects as used, and it does not free objects (destructor is
+// not called)
+template <class T>
+class SyncedPtrStableObjectList {
```
Comment (PENGUINLIONG) [inline taichi/rhi/impl_support.h:98]: so to my understanding the the contained objects should be explicitly constructed/destructed?

File: taichi/rhi/vulkan/vulkan_device.cpp [hunk 5]
```diff
@@ -1526,21 +1671,7 @@ DeviceAllocation VulkanDevice::allocate_memory(const AllocParams &params) {
...
-  assert(map_pair != allocations_.end() && "Invalid memory handle");
-  return map_pair->second;
-}
-
-const VulkanDevice::AllocationInternal &VulkanDevice::get_alloc_internal(
-    const DeviceAllocation &alloc) const {
-  const auto &map_pair = allocations_.find(alloc.alloc_id);
-  assert(map_pair != allocations_.cend() && "Invalid memory handle");
-  return map_pair->second;
+  return DeviceAllocation{this, (uint64_t)&alloc};
```
Comment (PENGUINLIONG) [inline taichi/rhi/vulkan/vulkan_device.cpp:1674]: finally!

**Review by PENGUINLIONG — APPROVED** [review]
lgtm! thank you

File: taichi/rhi/impl_support.h (hunk 4 above)
Comment (k-ye) [reply taichi/rhi/impl_support.h:98]: That's a good point. It seems like we should add a check that `T` is a POD-ish type, which doesn't have fancy ctor/dtor. (Dtor is especially important here, IMO. Right now the implementation would require the caller to call `~T()`)

Comment (bobcao3) [reply taichi/rhi/impl_support.h:98]: Fixed it now, the the object list now handles ctor/dtor & uses RAII.

File: taichi/rhi/vulkan/vulkan_device.cpp (hunk 3 above)
Comment (bobcao3) [reply taichi/rhi/vulkan/vulkan_device.cpp:860]: Should be doable. (really doubt there will be significant size difference tho)

File: taichi/rhi/impl_support.h (hunk 4 above)
Comment (k-ye) [reply taichi/rhi/impl_support.h:97]: nit: I suggest that we add some unit test on this (https://github.com/taichi-dev/taichi/tree/master/tests/cpp), given that this is a very low-level infra, and has certain behaviors that is not trivial (manually calling `~T()`, placement new, ...)

**Pull request merged**